### DISCUSSION TITLE
Wrapping Legacy AngularJS App as "Business logic" from Rails

### DISCUSSION THREAD

**Discussion by epugh** [rubycdp/ferrum#458] [Show and tell]
I wanted to share a spike using Ferrum for wrapping legacy Angular1 app to be run as a background job.  Quepid, https://github.com/o19s/quepid, is a tool for testing your search engine.  It has lots of custom code written in JS.   I'm using Ferrum to run a Chrome instance and execute the logic, return the JSON data, and then I can process that.   

Here is the in progress PR: https://github.com/o19s/quepid/pull/976

Specifically, I'd love feedback on how I am using Ferrum in my `run_case_job.rb`.   One issue I am noticing is that if I quit the browser via `browser.quit` then my Rails server quits!   It appears to quit ruby, bringing down Rails...?   So right now I have a `@@browser` object and just don't quit it.   Would love any suggestions, feedback, pointers to others who have used Ferrum in a Rails app!

Plus, tell me if I am crazy!

**Comment by route** [comment]
The better question would be how you run your jobs? The code doesn't tell us how you run it.

**Comment by epugh** [comment]
The jobs are run using ActiveJob, backed by SideKiq.   

My first attempt, with [run_case_job.rb](https://github.com/o19s/quepid/pull/976/files#diff-bec922e033742520978650641249b71b1edabf2e7361a5fbabea8de36c2220a5) instantiates Ferrum and then loads the web page that has my legacy angular app: `"http://localhost:3000/case/#{kase.id}/query/#{query.id}"`.  In that page is some JSON that is the data I need that I then pull out and use.   My second attempt, with [run_case_job2.rb](https://github.com/o19s/quepid/pull/976/files#diff-10f31ed017677fd5bb7d46b1919bd1de436caa529c79b4652dd77d990a650a28) instead calls a Rails request handler, [agent_q_controller.rb](https://github.com/o19s/quepid/pull/976/files#diff-e309b5534a7b249e7c41cd6c07a20b069722b6442dc0c4488529e9c5a1b1cede)  who then in turn instantiates Ferrum, calls the webpage with the legacy code, grabs the JSON data, and then returns that as the response from `agent_q_controller.rb` back to the calling ActiveJob `run_case_job2.rb`.  

The first attempt is obviously a lot less cumbersome however I wasn't sure if Ferrum would quit and bring down ActiveJob too!  I already had the `agent_q_controller.rb` so thought I would try that.

Right now however both ways periodically, say every 5 to 10 Ferrum calls, I then get a system exit, which then quits both my `web`  and my `worker` processes.

**Comment by epugh** [comment]
Oh, also, I *think* that calling `.quit` on the browser also kills my Rails process??!

**Reply by route** [reply]
Let me rephrase it. what command do you run to run rails server and/or jobs server? Is this procfile and heroku? What platform do you use?

**Reply by epugh** [reply]
Thank you for the interest in my challenges ;-).  Locally I run ferrum inside of a Docker container, and launch the web and worker processes using this procfile: https://github.com/o19s/quepid/blob/imagine_backend_batch/Procfile.dev.   I do install my own Chromium in my Docker container: https://github.com/o19s/quepid/blob/imagine_backend_batch/Dockerfile.dev#L24.

**Reply by route** [reply]
That means that rails is a dedicated process with it's own PID, as well as sidekiq. Calling quit in sidekiq by any means can't kill rails process. Check out your memory, maybe it's OOM.
